### DISCUSSION TITLE
What are the Axis for on a Fixed Constraint?

### DISCUSSION THREAD

**Discussion by DennisSmolek** [jrouwe/JoltPhysics.js#133] [Q&A] [answered]
In looking at [the docs for a Fixed Constraint](https://jrouwe.github.io/JoltPhysics/class_fixed_constraint_settings.html) it lists two sets of AxisX and AxisY values:
```
RVec3 | mPoint2 = RVec3::sZero()
Vec3 | mAxisX1 = Vec3::sAxisX()
Vec3 | mAxisY1 = Vec3::sAxisY()
RVec3 | mPoint2 = RVec3::sZero()
Vec3 | mAxisX2 = Vec3::sAxisX() 
Vec3 | mAxisY2 = Vec3::sAxisY()
```
I totally get what the `mPoint`'s are for, and I get what the Axis are for in the `HingeConstraint` but I can't seem to figure out what they are for in this fixed constraint..

**Comment by PhoenixIllusion** [comment, marked as answer]
Some of the constraint discussion was brought up on https://github.com/jrouwe/JoltPhysics.js/discussions/109

The 3x pairs of axis tie to both bodies Body1 (A) & Body2 (B), and any given X/Y/Z axis-set are required to be an orthogonal sets (90-degree from each other).

Once binding the bodies with the new Fixed Constraint, the engine will apply force to both bodies A&B to orient themselves so that their respective orthogonal sets align, regardless of what orientation they may have been on-start. You can always use the default world-space X/Y/Z axis and world-space fixed, and it will trigger "bind A & B with current relative orientation", but that might not be desired in some cases.

To just visibly define the significance of axis on this constraint, in a nonsense situation the below example, I have the blue ground, and a red pillar on it, bound at the magenta-point, and their constraint's Y-axis. If I was to fixed-constrain them together, but defined these particular Y-axis in the constraint, the physics system would forever apply the green force to attempt to force the pillar's Y-axis to align after activating the constraint. The base of the pillar won't allow it since it can't rotate into the ground, but the engine will keep trying.
![image](https://github.com/jrouwe/JoltPhysics.js/assets/756488/a1030ad8-be28-4784-b4f8-419744c171ec)

As a key part of these axis-definitions, you are defining a rule that regardless of what the orientation may be now or in the future (should they get dragged/pulled/forced/knocked out of alignment) they must attempt to reorient towards this defined target orientation with each other. In theory, on activating the constraint, the current orientation may not be exactly what you want, and possibly the world may never allow the bodies to ever reach the target orientation, but that is what you want to constrain to. Think if something got in the way of A&B before the constraint activated and jammed them in the wrong orientation.

For some of these, it helps to think of what's required to apply the constraint mid-way into a simulation. Some bodies are only supposed to bind together given certain exact orientations with each other that they might not be exactly at this-instant... Not exactly the following, but conceptually: tree trunks should only bind with their roots on the ground, car wheels only bind with the center of a wheel touching the axis of a wheel.

**Reply by DennisSmolek** [reply]
Very interesting! Thank you for the detailed reply!
